core_impl/src/main/java/com/google/instrumentation/trace/TraceExporterImpl.java

```diff
@@ -0,0 +1,184 @@
+    return new TraceExporterImpl(workerThread);
+    // TODO(bdrutu): Consider to add a shutdown hook to not avoid dropping data.
+  }
+
+  @Override
+  public void registerServiceHandler(String name, ServiceHandler serviceHandler) {
+    workerThread.registerServiceHandler(name, serviceHandler);
+  }
+
+  @Override
+  public void unregisterServiceHandler(String name) {
+    workerThread.unregisterServiceHandler(name);
+  }
+
+  /**
+   * Adds a {@link SpanImpl} to be exported.
+   *
+   * @param span the {@code SpanImpl} that will be exported.
+   */
+  void addSpan(SpanImpl span) {
+    workerThread.addSpan(span);
+  }
+
+  @VisibleForTesting
+  Thread getWorkerThread() {
+    return workerThread;
+  }
+
+  private TraceExporterImpl(WorkerThread workerThread) {
+    this.workerThread = workerThread;
+  }
+
+  // Worker thread that batches multiple span data and call the registered services to export
+  // these data.
+  //
+  // The map of registered handlers is implemented using ConcurrentHashMap ensuring full
+  // concurrency of retrievals and adjustable expected concurrency for updates, but retrievals
+  // reflect the results of the most recently completed update operations holding upon their onset.
+  //
+  // The list of batched data is protected by an explicit monitor object which ensures full
+  // concurrency.
+  private static final class WorkerThread extends Thread {
+    // Protects all variables.
+    private final Object monitor = new Object();
+
+    @GuardedBy("monitor")
+    private final List<SpanImpl> spans;
+
+    private final Map<String, ServiceHandler> serviceHandlers =
+        new ConcurrentHashMap<String, ServiceHandler>();
+    private final int bufferSize;
+    private final long scheduleDelayMillis;
+
+    // See TraceExporterImpl#addSpan.
+    private void addSpan(SpanImpl span) {
+      synchronized (monitor) {
+        this.spans.add(span);
+        if (spans.size() > bufferSize) {
+          monitor.notifyAll();
+        }
+      }
+    }
+
+    // See TraceExporterImpl#registerServiceHandler.
+    private void registerServiceHandler(String name, ServiceHandler serviceHandler) {
+      serviceHandlers.put(name, serviceHandler);
+    }
+
+    // See TraceExporterImpl#unregisterServiceHandler.
+    private void unregisterServiceHandler(String name) {
+      serviceHandlers.remove(name);
+    }
+
+    // Exports the list of SpanData to all the ServiceHandlers.
+    private void onBatchExport(List<SpanData> spanDataList) {
+      // From the java documentation of the ConcurrentHashMap#values():
+      // The view's iterator is a "weakly consistent" iterator that will never throw
+      // ConcurrentModificationException, and guarantees to traverse elements as they existed
+      // upon construction of the iterator, and may (but is not guaranteed to) reflect any
+      // modifications subsequent to construction.
+      for (Map.Entry<String, ServiceHandler> it : serviceHandlers.entrySet()) {
```

Comment:
here you could name the thread in a try-finally based on the service handler name. That way when blocking or other things happen, people can do a thread dump and notice what's going on. Ex. "TraceExporter("+it.getKey()+")"